src/main/scala/edu/berkeley/cs/rise/opaque/strategies.scala

```diff
@@ -136,25 +136,84 @@ object OpaqueOperators extends Strategy {
+            EncryptedProjectExec(resultExpressions,
+              EncryptedAggregateExec(groupingExpressions, aggregateExpressions, Some(Final),
+                EncryptedSortExec(groupingExpressions.map(_.toAttribute).map(e => SortOrder(e, Ascending)), true,
+                  EncryptedAggregateExec(groupingExpressions, aggregateExpressions, Some(Partial),
+                    EncryptedSortExec(groupingExpressions.map(e => SortOrder(e, Ascending)), false, planLater(child)))))) :: Nil
+          }
+        case size if size == 1 => // One distinct aggregate operation
+          // Because we are also grouping on the columns used in the distinct expressions,
+          // we do not need separate cases for global and grouping aggregation.
+
+          val namedDistinctExpressions = functionsWithDistinct.head.aggregateFunction.children.map { e =>
+            e match {
+              case ne: NamedExpression => ne
+              case other =>
+                // Keep the name of the original expression.
+                val name = e match {
+                  case ne: NamedExpression => ne.name
+                  case _ => e.toString
+                }
+                Alias(other, name)()
+            }
+          }
+          val combinedGroupingExpressions = groupingExpressions ++ namedDistinctExpressions
+
+          // 1. Create an Aggregate operator for partial aggregations.
+          val partialAggregate = {
+            val sorted = EncryptedSortExec(combinedGroupingExpressions.map(e => SortOrder(e, Ascending)), false,
+              planLater(child))
+            EncryptedAggregateExec(combinedGroupingExpressions, functionsWithoutDistinct, Some(Partial), sorted)
+          }
+
+          // 2. Create an Aggregate operator for partial merge aggregations.
+          val partialMergeAggregate = {
+            val sorted = EncryptedSortExec(combinedGroupingExpressions.map(e => SortOrder(e, Ascending)), true,
+              partialAggregate)
+            EncryptedAggregateExec(combinedGroupingExpressions,
+              functionsWithoutDistinct, Some(PartialMerge), sorted)
+          }
+
+          // 3. Create an Aggregate operator for partial aggregation of distinct aggregate expressions.
+          val partialDistinctAggregate = {
+            // Indistinct functions operate on aggregation buffers since partial aggregation was already called,
+            // but distinct functions operate on the original input to the aggregation.
+            EncryptedAggregateExec(groupingExpressions,
+              functionsWithoutDistinct.map(_.copy(mode = PartialMerge)) ++
+                functionsWithDistinct.map(_.copy(mode = Partial)), None, partialMergeAggregate)
+          }
+
+          // 4. Create an Aggregate operator for the final aggregation.
+          val finalAggregate = {
+            val sorted = EncryptedSortExec(groupingExpressions.map(e => SortOrder(e, Ascending)),
+              true, partialDistinctAggregate)
```

Comment:
Unfortunately, I could not find a way to get rid of this global sort. Without it, the result is off by the following margin: `Set([B,4,11], [C,4,10], [A,4,11]) did not equal Set([C,1,2], [B,3,8], [B,1,3], [C,3,8], [A,4,11])`

Reply:
I think it's because in the first global sort (in the second step), the sort is executed on `combinedGroupingExpressions`? This does not guarantee that the rows belonging to the same group under `groupingExpressions` only will be colocated to the same partition. You can solve it by first doing an explicit range partitioning on `groupingExpressions`, followed by a per partition sort on `combinedGroupingExpressions`. A similar strategy is currently used in join: https://github.com/mc2-project/opaque/blob/master/src/main/scala/edu/berkeley/cs/rise/opaque/strategies.scala#L92.